### PR TITLE
feat: deterministic budget-pressure notices for session self-paging

### LINKED ISSUE #1216: Deterministic context-pressure notices (session self-paging, no manager model)
# Deterministic context-pressure notices (session self-paging, no manager model)

## Problem

Between "everything fine" and "compaction fired" there is no signal. A
session grows from 0% to 80% with the model blind to its budget state, then
compaction summarizes everything at once — including facts the model would
have banked into memory had it known pressure was building. MemGPT's
transferable insight (arXiv 2310.08560) is that the *session model itself*
can page memory via tools when driven by deterministic budget warnings; no
manager model is required.

jcode already owns every prerequisite: per-turn token accounting
(`context_usage_with`), the `memory` tool (remember/recall/search), and the
transient `<system-reminder>` injection. Only the signal is missing.

## Proposal

Two usage bands with hysteresis, computed from existing accounting in
`messages_for_provider`, delivered through the existing transient reminder
(which never touches persisted history):

- 75% advisory: bank durable facts/decisions (`memory` remember), recall on
  demand instead of keeping everything live. Explicitly no action needed if
  remaining work is short.
- 90% urgent: bank NOW — after compaction only the summary survives.
- Bands fire once each and re-arm below 70%, so a turn hovering at a
  threshold notifies once, not every turn. No notice fires on a turn where
  compaction itself applied (that path already reports).

Constants (`PRESSURE_ADVISORY/URGENT/REARM`), a pure decision function, and
~20 lines of wiring. Zero behavior change below 70% usage. The reminder
names only tools that exist, so the model can act immediately.

## Reference implementation

Built and verified on `origin/master` (983b532). One commit on the fork:

- `SK-DEV-AI/jcode@807acca` — the feature + tests

Full diff:

https://github.com/SK-DEV-AI/jcode/compare/1jehuang:master...SK-DEV-AI:feat/pressure-notices

## Verification

- 8 band-edge cases on the pure decision function (thresholds, no re-fire
  within a band, escalation 1→2, silence below/above).
- Integration test through `messages_for_provider` with a 1000-token budget:
  fires once at ~80% with tool-grounded text, stays silent on the next turn
  at the same pressure, re-fires after relief (budget raised, band re-armed)
  and regrowth.
- Full `jcode-app-core` suite baselined against clean `origin/master`:
  identical pre-existing failure set, zero regressions. `cargo fmt` clean
  (per-file, no unrelated normalization), no new clippy warnings.

Happy for this to land rewritten in your own shape, as with #1174.

### PR DESCRIPTION
**Pull request by SK-DEV-AI**
Closes #1216 as built (design agreed in thread; opening the code for review rather than waiting on another gate round).

Pure decision function plus ~30 lines into the existing accounting and transient-reminder path: 75% advisory, 90% urgent, re-arm below 70%, zero behavior change below 70%. Text names only real memory tools so the session can bank now and recall later.

Tests: 9 edge cases on the decision function plus integration through messages_for_provider proving fires-once, no-duplicate, re-fire-after-relief. Full app-core suite zero-regression vs clean tree (pre-existing failures identical, verified by stash comparison).

### CODE REVIEW

File: crates/jcode-app-core/src/agent.rs [hunk 1]
```diff
@@ -761,6 +809,27 @@ impl Agent {
...
+                    if usage < Self::PRESSURE_REARM {
+                        self.last_pressure_band = 0;
+                    } else if event.is_none() {
+                        if let Some(band) =
+                            Self::pressure_band_for_usage(usage, self.last_pressure_band)
+                        {
+                            self.last_pressure_band = band;
+                            let tokens = manager.effective_token_count_with(&messages) as u64;
+                            self.current_turn_system_reminder =
+                                Some(Self::pressure_notice(band, usage * 100.0, tokens));
```
[the remark this answers is not in the document]
Comment (SK-DEV-AI) [reply crates/jcode-app-core/src/agent.rs]: Fixed. messages_for_provider now raises a fired flag and both turn loops rebuild the already-built prompt when it fires without a compaction event (compaction already rebuilds; static part untouched so prewarm stays valid). The crossing turn carries the notice; the flag lowers on non-firing turns so it never repeats or rebuilds needlessly.

File: crates/jcode-app-core/src/agent.rs [hunk 2]
```diff
@@ -206,7 +206,13 @@ pub struct Agent {
     pending_alerts: Vec<String>,
     /// Transient reminder injected into provider requests for the current turn only.
     /// Not persisted to session history.
-    current_turn_system_reminder: Option<String>,
+    pub(crate) current_turn_system_reminder: Option<String>,
+    /// Highest context-pressure band already notified this session cycle (0 =
+    /// none, 1 = advisory, 2 = urgent). Resets when usage drops back below
+    /// the re-arm level (e.g. after compaction) so notices can fire again on
+    /// regrowth. Backs the deterministic budget notices: the session model
+    /// pages its own memory instead of needing a manager model.
```
[the remark this answers is not in the document]
Comment (SK-DEV-AI) [reply crates/jcode-app-core/src/agent.rs:214]: Fixed. last_pressure_band (and the fired flag) reset in reset_runtime_state_for_session_change, so a restored session above the old band gets its own notice. Covered by a restore test.

File: crates/jcode-app-core/src/agent.rs [hunk 3]
```diff
@@ -728,7 +735,48 @@ impl Agent {
...
+                 after compaction only the summary survives. Recall (recall/search) on demand instead of keeping everything live.",
+            ),
+            _ => format!(
+                "Context is {usage_pct:.0}% full (~{tokens} tokens). Before continuing, consider \
+                 banking durable facts and decisions with the memory tool (remember) and recalling \
+                 (recall/search) only what is needed, so the coming compaction has less to summarize. \
+                 No action needed if the remaining work is short.",
+            ),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (SK-DEV-AI) [reply crates/jcode-app-core/src/agent.rs]: Fixed. The notice names the memory tool only when the session can call it (agent allow/disabled sets plus the session policy via a new session_tool_policy_allows_tool query); restricted sessions get tool-neutral banking language. Covered by a restricted-agent test.

File: crates/jcode-app-core/src/agent.rs [hunk 4]
```diff
@@ -728,7 +744,80 @@ impl Agent {
...
+            "Recall (recall/search) on demand instead of keeping everything live."
+        } else {
+            "Keep only what is needed live."
+        };
+        match band {
+            2 => format!(
+                "Context is {usage_pct:.0}% full (~{tokens} tokens) and compaction is near. \
+                 {bank_line} — after compaction only the summary survives, stale tool output \
+                 is dropped automatically, and middle context is already unreliable: do not \
+                 trust details you have not banked or re-read. {recall_line}",
```
[the remark this answers is not in the document]
Comment (SK-DEV-AI) [reply crates/jcode-app-core/src/agent.rs]: Fixed. The urgent text now says only the summary plus the most recent messages survive — the retained tail (including recent tool results) is no longer misdescribed as dropped.
